pmd-core/src/main/java/net/sourceforge/pmd/properties/AbstractPropertySource.java

```diff
@@ -58,24 +92,20 @@
     @Override
     public void definePropertyDescriptor(PropertyDescriptor<?> propertyDescriptor) {
         // Check to ensure the property does not already exist.
-        for (PropertyDescriptor<?> descriptor : propertyDescriptors) {
-            if (descriptor.name().equals(propertyDescriptor.name())) {
-                throw new IllegalArgumentException("There is already a PropertyDescriptor with name '"
-                                                   + propertyDescriptor.name() + "' defined on Rule " + getName() + ".");
-            }
+        if (getPropertyDescriptor(propertyDescriptor.name()) != null) {
+            throw new IllegalArgumentException("There is already a PropertyDescriptor with name '"
+                                                       + propertyDescriptor.name() + "' defined on " + getPropertySourceType() + " " + getName() + ".");
+
         }
         propertyDescriptors.add(propertyDescriptor);
         // Sort in UI order
         Collections.sort(propertyDescriptors);
     }
 
 
-    /**
-     * Gets the name of the property source. This is e.g. the rule name or the report name.
-     *
-     * @return the name
-     */
-    public abstract String getName();
+    private String getPropertySourceType() {
```

Comment:
We probably should make this protected and abstract, so that the AbstractPropertySource doesn't need to know its concrete subclasses... like Rule.

Reply:
I support this